<a href="https://colab.research.google.com/github/VivekSalunkhe14/Cascade/blob/main/Cascade_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# **Importing Libraries**

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import calendar

# **Dataset for Analysis of absenteeism in a company** 


The dataset was created with records of **Absenteeism at Work** from July 2007 to July 2010 at a courier company in Brazil.
Attribute Information of the dataset is as follows:
1. Individual identification (ID)
2. Reason for absence (ICD)
3. Month of absence
4. Day of the week (Monday (2), Tuesday (3), Wednesday (4), Thursday (5), Friday (6))
5. Seasons (summer (1), autumn (2), winter (3), spring (4))
6. Transportation expense
7. Distance from Residence to Work (kilometers)
8. Service time
9. Age
10. Work load Average/day
11. Hit target
12. Disciplinary failure (yes=1; no=0)
13. Education (high school (1), graduate (2), postgraduate (3), master and doctor (4))
14. Son (number of children)
15. Social drinker (yes=1; no=0)
16. Social smoker (yes=1; no=0)
17. Pet (number of pet)
18. Weight
19. Height
20. Body mass index
21. Absenteeism time in hours (target)

In [ ]:
absent = pd.read_csv('/content/drive/MyDrive/Absenteeism_at_work.csv')

In [ ]:
absent.head()

ID  Reason for absence  ...  Body mass index  Absenteeism time in hours
0  11                  26  ...               30                          4
1  36                   0  ...               31                          0
2   3                  23  ...               31                          2
3   7                   7  ...               24                          4
4  11                  23  ...               30                          2

[5 rows x 21 columns]

In [ ]:
absent.shape

(740, 21)

In [ ]:
absent.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 740 entries, 0 to 739
Data columns (total 21 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   ID                               740 non-null    int64  
 1   Reason for absence               740 non-null    int64  
 2   Month of absence                 740 non-null    int64  
 3   Day of the week                  740 non-null    int64  
 4   Seasons                          740 non-null    int64  
 5   Transportation expense           740 non-null    int64  
 6   Distance from Residence to Work  740 non-null    int64  
 7   Service time                     740 non-null    int64  
 8   Age                              740 non-null    int64  
 9   Work load Average/day            740 non-null    float64
 10  Hit target                       740 non-null    int64  
 11  Disciplinary failure             740 non-null    int64  
 12  Education             

The dimension of the dataset clearly indicates that our analysis will be focused on 21 features and 740 datapoints. We observed that the data type of all the features is integer which is incorrect as there are few columns with categorical data and their data type should be appropriately specified as category 

#**Assigning appropriate data types to columns**

In [ ]:
absent['Reason for absence'] = absent['Reason for absence'].astype('category')
absent['Month of absence'] = absent['Month of absence'].astype('category')
absent['Day of the week'] = absent['Day of the week'].astype('category')
absent['Seasons'] = absent['Seasons'].astype('category')
absent['Education'] = absent['Education'].astype('category')
absent['Disciplinary failure'] = absent['Disciplinary failure'].astype('category')
absent['Social drinker'] = absent['Social drinker'].astype('category')  
absent['Social smoker'] = absent['Social smoker'].astype('category') 

In [ ]:
absent.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 740 entries, 0 to 739
Data columns (total 21 columns):
 #   Column                           Non-Null Count  Dtype   
---  ------                           --------------  -----   
 0   ID                               740 non-null    int64   
 1   Reason for absence               740 non-null    category
 2   Month of absence                 740 non-null    category
 3   Day of the week                  740 non-null    category
 4   Seasons                          740 non-null    category
 5   Transportation expense           740 non-null    int64   
 6   Distance from Residence to Work  740 non-null    int64   
 7   Service time                     740 non-null    int64   
 8   Age                              740 non-null    int64   
 9   Work load Average/day            740 non-null    float64 
 10  Hit target                       740 non-null    int64   
 11  Disciplinary failure             740 non-null    category
 12  Educatio

After assigning proper datatype to the categorical columns we can observe optimization in space as storage size dropped to 84.1 KB from 121.5 KB. Once the columns were of proper type it was mandatory to clean the data for our efficient use.

In [ ]:
absent.isnull().sum()

ID                                 0
Reason for absence                 0
Month of absence                   0
Day of the week                    0
Seasons                            0
Transportation expense             0
Distance from Residence to Work    0
Service time                       0
Age                                0
Work load Average/day              0
Hit target                         0
Disciplinary failure               0
Education                          0
Son                                0
Social drinker                     0
Social smoker                      0
Pet                                0
Weight                             0
Height                             0
Body mass index                    0
Absenteeism time in hours          0
dtype: int64

After verifying thst there were no NA values in the dataset our next step was verification of valid datapoints in the dataset

As mentioned in dataset description the feature named **Reason for absence** can take any value between 1 to 28, however 0 value for the Reason which can be a human error was replaced with 26 (as it stands for **Unjustified Abscence** in data description.)

In [ ]:
len(absent[absent['Reason for absence']==0])

43

In [ ]:
absent.loc[(absent['Reason for absence']==0),'Reason for absence']=26

In [ ]:
len(absent[absent['Reason for absence']==0])

0

#**Mapping value of character features to its pre-defined categories.**
Since our dataset had all the columns as numeric it was necessary to create few character features from the categorical columns for easy interpretation of the data.
With the help of data description the categorical features namely **Reason for absence**, **Month of absence**, **Day of the week**, **Season**, **Education**, **Disciplinary failure**, **Social drinker**, **Social smoker** were mapped to its understandable character features or columns.

Features obtained after this preprocessing will be very benefical during Exploratory Data Analysis.

In [ ]:
reason_dict = {
    1: 'Certain infectious and parasitic diseases',
    2: 'Neoplasms',
    3: 'Diseases of the blood and blood-forming organs and certain disorders involving the immune mechanism',
    4: 'Endocrine, nutritional and metabolic diseases',
    5: 'Mental and behavioural disorders',
    6: 'Diseases of the nervous system',
    7: 'Diseases of the eye and adnexa',
    8: 'Diseases of the ear and mastoid process',
    9: 'Diseases of the circulatory system',
    10: 'Diseases of the respiratory system',
    11: 'Diseases of the digestive system',
    12: 'Diseases of the skin and subcutaneous tissue',
    13: 'Diseases of the musculoskeletal system and connective tissue',
    14: 'Diseases of the genitourinary system',
    15: 'Pregnancy, childbirth and the puerperium',
    16: 'Certain conditions originating in the perinatal period',
    17: 'Congenital malformations, deformations and chromosomal abnormalities',
    18: 'Symptoms, signs and abnormal clinical and laboratory findings, not elsewhere classified',
    19: 'Injury, poisoning and certain other consequences of external causes',
    20: 'External causes of morbidity and mortality',
    21: 'Factors influencing health status and contact with health services',
    22: 'Patient follow-up',
    23: 'Medical consultation',
    24: 'Blood donation',
    25: 'Laboratory examination',
    26: 'Unjustified absence',
    27: 'Physiotherapy',
    28: 'Dental consultation'
}
absent['Reason_Justified'] = absent['Reason for absence'].map(reason_dict).astype('category')

In [ ]:
absent['Month_Name'] =  absent['Month of absence'].apply(lambda x: calendar.month_abbr[x]).astype('category')

In [ ]:
day_dict = {
    2: 'Monday',
    3: 'Tuesday',
    4: 'Wednesday',
    5: 'Thursday',
    6: 'Friday',
}
absent['Day_Name'] = absent['Day of the week'].map(day_dict).astype('category')

In [ ]:
season_dict = {
    1:'Summer', 
    2:'Autumn', 
    3:'Winter',
    4:'Spring'
}
absent['Season_Name'] = absent.Seasons.map(season_dict).astype('category')

In [ ]:
education_dict = {
    1: 'High School',
    2: 'Graduate',
    3: 'Post Graduate',
    4: 'Master & Doctor'
}
absent['Education_Details'] = absent['Education'].map(education_dict).astype('category')

In [ ]:
discipline_dict = {
    0: 'Disciplined',
    1: 'Indisciplined',
}
absent['Disciplinary_Status'] = absent['Disciplinary failure'].map(discipline_dict).astype('category')

In [ ]:
drinker_dict = {
    0: 'Nondrinker',
    1: 'Drinker',
}
absent['Drinking_Status'] = absent['Social drinker'].map(drinker_dict).astype('category')

In [ ]:
smoker_dict = {
    0: 'Nonsmoker',
    1: 'Smoker',
}
absent['Smoking_Status'] = absent['Social smoker'].map(smoker_dict).astype('category')

In [ ]:
absent.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 740 entries, 0 to 739
Data columns (total 29 columns):
 #   Column                           Non-Null Count  Dtype   
---  ------                           --------------  -----   
 0   ID                               740 non-null    int64   
 1   Reason for absence               740 non-null    category
 2   Month of absence                 740 non-null    category
 3   Day of the week                  740 non-null    category
 4   Seasons                          740 non-null    category
 5   Transportation expense           740 non-null    int64   
 6   Distance from Residence to Work  740 non-null    int64   
 7   Service time                     740 non-null    int64   
 8   Age                              740 non-null    int64   
 9   Work load Average/day            740 non-null    float64 
 10  Hit target                       740 non-null    int64   
 11  Disciplinary failure             740 non-null    category
 12  Educatio

# **Inconsistency in Month column**

Feature **Month of absence** can take any value between 1 to 12 as it suggests month number for a particular year but 0 value clearly dictates error in datapoints which need to be dropped.

In [ ]:
absent[absent['Month of absence'] == 0]

ID Reason for absence  ... Drinking_Status Smoking_Status
737   4                 26  ...         Drinker      Nonsmoker
738   8                 26  ...         Drinker      Nonsmoker
739  35                 26  ...      Nondrinker      Nonsmoker

[3 rows x 29 columns]

In [ ]:
absent = absent[absent['Month of absence'] != 0]

In [ ]:
absent.shape

(737, 29)

# **Inconsistency in ID column**

Initially we observed that there were 34 Unique ID's in the dataset


In [ ]:
len(absent['ID'].unique())

34

However after grouping the datasets based on some essential features we observed that there was discrepancy in count of ID's and one of the idea was duplicated

In [ ]:
sum(absent.groupby(['Smoking_Status','Drinking_Status','Age','Body mass index'])['ID'].nunique())

35

Dataframe with all the 34 Unique ID's in dataset

In [ ]:
df1 = pd.DataFrame({'ID': absent['ID'].unique()})

In order to identify the extra duplicated ID record we grouped the dataset by ID and counted the unique values of Age. Once this was done we extracted the recorded containing count of ID greater than 1. We observed that there was an ID 29 present with count of 2 which clearly indicates that one of the record with ID 29 is incorrect since all the similar ID will have same Age as they indicate the same individual .

In [ ]:
df3 = absent.groupby('ID')['Age'].nunique().reset_index(name='Count of ID')
df3[df3['Count of ID']>1]

ID  Count of ID
27  29            2

In [ ]:
absent.loc[absent['ID']==29,['ID','Age','Weight','Height','Drinking_Status','Smoking_Status']]

ID  Age  Weight  Height Drinking_Status Smoking_Status
51   29   28      69     169      Nondrinker      Nonsmoker
592  29   41      94     182         Drinker      Nonsmoker
675  29   41      94     182         Drinker      Nonsmoker
681  29   41      94     182         Drinker      Nonsmoker
683  29   41      94     182         Drinker      Nonsmoker

After extracting the data of 29th ID we found out that there was one record with Age 28 which seems incorrect as all the other datapoints were having Age 41 for 29th ID. We then drop that particular inconsistent record from our dataset

In [ ]:
absent = absent.drop(51)
absent.shape

(736, 29)

In [ ]:
absent.loc[absent['ID']==29,['ID','Age','Weight','Height','Drinking_Status','Smoking_Status']]

ID  Age  Weight  Height Drinking_Status Smoking_Status
592  29   41      94     182         Drinker      Nonsmoker
675  29   41      94     182         Drinker      Nonsmoker
681  29   41      94     182         Drinker      Nonsmoker
683  29   41      94     182         Drinker      Nonsmoker

# **Inconsistency in Absenteeism time in hours column**

Dataset aims to provide information for absenteeism but, 0 value in "Absenteeism time in hours" indicates discrepancy in data and we need to be specify certain value for such records as per constraints to remove contradiction.

In [ ]:
len(absent[absent['Absenteeism time in hours'] == 0])

40

In [ ]:
len(absent[absent['Disciplinary_Status'] == 'Indisciplined'])

39

There exist 40 rows for 0 absenteeism time but only 39 rows for Indisciplinary status. One row must be valid which states that ID is disciplined and having valid reason. Let us identify that particular valid row.

In [ ]:
absent[(absent['Absenteeism time in hours']==0) & (absent['Disciplinary_Status']!='Indisciplined')]

ID Reason for absence  ... Drinking_Status Smoking_Status
134  34                 27  ...      Nondrinker      Nonsmoker

[1 rows x 29 columns]

For this particular one datapoint the Absenteeism time was replaced with average value of absent time for the reason stated by  that corresponding ID i.e. due to Physiotherapy.

In [ ]:
absent[absent['Reason for absence']==27][['Absenteeism time in hours']].mean()

Absenteeism time in hours    2.275362
dtype: float64

In [ ]:
absent.loc[(absent['Reason for absence']==27) & (absent['Absenteeism time in hours']==0),'Absenteeism time in hours']=2

In [ ]:
len(absent[(absent['Absenteeism time in hours']==0) & (absent['Disciplinary_Status']!='Indisciplined')])

0

After replacing that particular ID we are left with 39 datapoints whose absenteeism time was left as 0

In [ ]:
len(absent[absent['Absenteeism time in hours']==0])

39

It is assumed that absenteeism time for those 39 data points will be mode value of all Disciplined data absenteeism time. So, 8 hours is assigned for proper EDA analysis.

In [ ]:
(absent[absent['Disciplinary_Status']=='Disciplined'][['Absenteeism time in hours']]).mode()

Absenteeism time in hours
0                          8

In [ ]:
absent.loc[(absent['Absenteeism time in hours']==0),'Absenteeism time in hours']=8

In [ ]:
len(absent[absent['Absenteeism time in hours']==0])

0

After removing all inconsistency we are left with 0 records with absenteeism time having 0 value which indicates dataset now contains valid data

# **Extracting useful information by using calculation of features.**

Travel distance and travel cost used to vary in large range which will be difficult to compare and conclude useful information so Travel Expense per kilometer (R$/km) will help to compare expense by any ID per km to industry.

In [ ]:
absent['Travel Expense per km'] = round((absent['Transportation expense']/absent['Distance from Residence to Work']),2)

In [ ]:
absent.head()

ID Reason for absence  ... Smoking_Status Travel Expense per km
0  11                 26  ...      Nonsmoker                  8.03
1  36                 26  ...      Nonsmoker                  9.08
2   3                 23  ...      Nonsmoker                  3.51
3   7                  7  ...         Smoker                 55.80
4  11                 23  ...      Nonsmoker                  8.03

[5 rows x 30 columns]

Creating categorical column **Travel cost category** for specifying which travel group the individual belongs.

In [ ]:
absent.loc[absent['Travel Expense per km']<7, 'Travel Cost Category'] = 'Cheap'
absent.loc[absent['Travel Expense per km']>15, 'Travel Cost Category'] = 'Expensive'
absent.loc[(absent['Travel Expense per km']>=7) & (absent['Travel Expense per km']<=15), 'Travel Cost Category'] = 'Affordable'

In [ ]:
absent.head()

ID Reason for absence  ... Travel Expense per km Travel Cost Category
0  11                 26  ...                  8.03           Affordable
1  36                 26  ...                  9.08           Affordable
2   3                 23  ...                  3.51                Cheap
3   7                  7  ...                 55.80            Expensive
4  11                 23  ...                  8.03           Affordable

[5 rows x 31 columns]

A new feature **Joining_age** is created which will provide information about the employee's joining status. Employee having high value of joining age will definitely have good working experience and also pose a chance that particular individual will have higher educational qualifications than others.

In [ ]:
absent['Joining_Age'] = absent['Age'] - absent['Service time']

In [ ]:
absent.head()

ID Reason for absence  ... Travel Cost Category Joining_Age
0  11                 26  ...           Affordable          20
1  36                 26  ...           Affordable          32
2   3                 23  ...                Cheap          20
3   7                  7  ...            Expensive          25
4  11                 23  ...           Affordable          20

[5 rows x 32 columns]

Age feature is classified into three categories as young, mid-age and old and further analysis could be carried out for each age category and important parameters affecting each age category can also be obtained.

In [ ]:
absent.loc[absent['Age']<35, 'Age_Group'] = 'Young Employee'
absent.loc[absent['Age']>45, 'Age_Group'] = 'Old Employee'
absent.loc[(absent['Age']>=35) & (absent['Age']<=45), 'Age_Group'] = 'Mid-Age Employee'

In [ ]:
absent.head()

ID Reason for absence  ... Joining_Age         Age_Group
0  11                 26  ...          20    Young Employee
1  36                 26  ...          32      Old Employee
2   3                 23  ...          20  Mid-Age Employee
3   7                  7  ...          25  Mid-Age Employee
4  11                 23  ...          20    Young Employee

[5 rows x 33 columns]

Body mass index feature is specified for each individual which is used to create a categorical feature **BMI_Status** with 3 categories i.e Normal, Overweight and Obese. This categorical feature will further provide efficient way to perform EDA on the data.

In [ ]:
absent.loc[absent['Body mass index']<18.5, 'BMI_Status'] = 'Underweight'
absent.loc[(absent['Body mass index']>=18.5) & (absent['Body mass index']<24.9), 'BMI_Status'] = 'Normal'
absent.loc[(absent['Body mass index']>=24.9) & (absent['Body mass index']<29.9), 'BMI_Status'] = 'Overweight'
absent.loc[absent['Body mass index']>29.9, 'BMI_Status'] = 'Obese'

In [ ]:
absent.head()

ID Reason for absence  ...         Age_Group BMI_Status
0  11                 26  ...    Young Employee      Obese
1  36                 26  ...      Old Employee      Obese
2   3                 23  ...  Mid-Age Employee      Obese
3   7                  7  ...  Mid-Age Employee     Normal
4  11                 23  ...    Young Employee      Obese

[5 rows x 34 columns]

# **Output CSV file for performing Exploratory Data Analysis**

In [ ]:
absent.to_csv('/content/drive/MyDrive/Absent_Preprocessed.csv',index=False)